File: src/main/golang/leetcode/editor/cn/[187]Repeated DNA Sequences.go
package lc

import "math"

//leetcode submit region begin(Prohibit modification and deletion)
func findRepeatedDnaSequences(s string) (ans []string) {
	getNum := func(char rune) (num int) {
		switch char {
		case 'A':
			num = 0
		case 'C':
			num = 1
		case 'G':
			num = 2
		case 'T':
			num = 3
		}
		return
	}
	R, L := 4, 10
	RL := int(math.Pow(float64(R), float64(L-1)))
	var whash int
	visited := make(map[int]int)
	for i, char := range s {
		num := getNum(char)
		whash = R*whash + num
		if i >= 9 {
			if v, _ := visited[whash]; v == 1 {
				ans = append(ans, s[i-9:i+1])
			}
			visited[whash] += 1
			whash -= getNum(rune(s[i-9])) * RL
		}
	}
	return
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[187]Repeated DNA Sequences.py
from typing import List
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        if len(s) <= 10:
            return []
        res = []
        dic = {}
        for i in range(len(s) - 9):
            if s[i:i+10] not in dic:
                dic[s[i:i+10]] = 1
            else:
                dic[s[i:i+10]] += 1
                if dic[s[i:i+10]] == 2:
                    res.append(s[i:i+10])
        return res
# leetcode submit region end(Prohibit modification and deletion)
